python/cudf/cudf/core/frame.py

```diff
@@ -321,15 +321,19 @@ def __len__(self):
 
     @_cudf_nvtx_annotate
     def astype(self, dtype, copy=False, **kwargs):
-        result = {}
+        result_data = {}
         for col_name, col in self._data.items():
             dt = dtype.get(col_name, col.dtype)
             if not is_dtype_equal(dt, col.dtype):
-                result[col_name] = col.astype(dt, copy=copy, **kwargs)
+                result_data[col_name] = col.astype(dt, copy=copy, **kwargs)
             else:
-                result[col_name] = col.copy() if copy else col
+                result_data[col_name] = col.copy() if copy else col
 
-        return result
+        return ColumnAccessor._create_unsafe(
+            data=result_data,
+            multiindex=self._data.multiindex,
+            level_names=self._data.level_names,
+        )
```

Comment:
Why does `Frame.astype` return a `ColumnAccessor`? Should this not return `self._from_data(data=ColumnAccessor)` ?

Reply:
@vyasr was this a conscious decision to return dict here in: https://github.com/rapidsai/cudf/pull/10381?

Reply:
Think this is because `IndexedFrame.astype` calls this to build the `data` argument to `_from_data`. Since that just takes a `MutableMapping` either worked I guess, though previously multiindices were not passed in?

Reply:
It was a conscious decision because `IndexedFrame.astype` and `GenericIndex.astype` both use the returned value to construct the relevant objects, but in the case of `GenericIndex` we have to construct the appropriate subclass which means we have to use `_index_from_data` rather than `self._from_data`. Ideally `Frame.astype` would take the `data` it's currently returning and instead return `self._from_data(data)`, but that makes the code harder to reuse in `GenericIndex` and forces extra creation of cudf objects so I chose not to do that. In hindsight, it would probably have been better to rename the method to `Frame._astype` or something similar to indicate that it's a helper function that doesn't return exactly the same thing as all the public `astype` APIs. I would support such a rename, or a restructuring of the code if you could find a solution that reduced duplication further in the subclasses.

Reply:
I'd be +1 for renaming to the internal method `Frame._astype`. I can't think of a better refactor than it is currently, and don't want to make a widespread refactor at end of a release. WDYT @wence- ?

Reply:
Let's raise an issue for 23.0x and leave as is here in the interrsts of making progress

Reply:
done: https://github.com/rapidsai/cudf/issues/12140